BasicSample/Assets/GrabAndThrow/Scripts/GrabAndThrowBehavior.cs

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using UnityEngine;
+using UnityEngine.XR;
+
+namespace Microsoft.MixedReality.OpenXR.BasicSample
+{
+	[RequireComponent(typeof(Rigidbody))]
+	public class GrabAndThrowBehavior : MonoBehaviour
+	{
+		private const float PICKABLE_RANGE = 0.5f;
+
+		[SerializeField]
+		public BoxCollider PickupDetectionCollider;
+
+		private bool[] m_wasDeviceTapped = { false, false };
+		private Rigidbody rigidBody;
+
+		private XRNode? handHeldBy = null;
+		private Transform handSpace;
+		private Vector3 positionInHandSpace;
+		private Quaternion rotationInHandSpace;
+
+		private void Awake()
+		{
+			rigidBody = GetComponent<Rigidbody>();
+			GameObject handSpaceGameObject = new GameObject("Hand Space");
+			handSpace = Instantiate(handSpaceGameObject, Vector3.zero, Quaternion.identity).transform;
+		}
+
+		// Update is called once per frame
+		private void Update()
```

Comment:
If I understand this correctly, you are attaching this behavior to every objects that can be picked up.   This will introduce a lot virtual function calls on each object and the "update" logic probably running very similar code to read data from input device, repeated for every objects.

Is it possible to do this the other way.  Give those pickable objects a tag or something so that you can run this "pick up logic" just once per hand per frame and loop through the objects to detect which object to pickup?

Reply:
Refactored this to use GrabAndThrowManager.cs for handling update logic, with ThrowableObject.cs used to tag various objects valid for interaction.